### PR TITLE
net: lwm2m: LwM2M Portfolio object id support

### PR DESCRIPTION
**Pull request by jheiskan81**
Added support for Portfolio object support because LwM2M v1.1 conformance
test requirement that. This object is only for conformance test purposing.

### CODE REVIEW

**Comment by jheiskan81** [comment]
@SeppoTakalo @VeijoPesonen @rlubos @jarlamsa Please review.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * add PORTFOLIO_IDENTITY_ID resource instances
+ * add PORTFOLIO_AUTH_DATA_ID resource instances
+ */
+#define RESOURCE_INSTANCE_COUNT        (PORTFOLIO_MAX_ID - 3 + \
+					PORTFOLIO_IDENTITY_MAX + \
+					PORTFOLIO_AUTH_DATA_MAX)
+
+static struct lwm2m_engine_obj portfolio;
+static struct lwm2m_engine_obj_field fields[] = {
+	OBJ_FIELD_DATA(PORTFOLIO_IDENTITY_ID, RW, STRING), /* Mandaory Multi Instance */
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:49]: Typo, `OBJ_FIELD_DATA(PORTFOLIO_IDENTITY_ID, RW, STRING), /* Mandatory Multi Instance */`

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	OBJ_FIELD_EXECUTE(PORTFOLIO_GET_AUTH_DATA_ID), /* Optional */
+	OBJ_FIELD_DATA(PORTFOLIO_AUTH_DATA_ID, R_OPT, OPAQUE), /* Optional Multi instance */
+	OBJ_FIELD_DATA(PORTFOLIO_AUTH_STATUS_ID, R_OPT, U8), /* Optional: Single */
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:52]: Drop the colon.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+			LOG_ERR("Can not create instance - "
+				"already existing: %u",
+				obj_inst_id);
+			return NULL;
+		}
+
+		/* Save first available slot index */
+		if (avail < 0 && !inst[index].obj) {
+			avail = index;
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:74]: `break;` ?

**Review by VeijoPesonen — COMMENTED** [review]
Minor style issues plus a question.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c (hunk 1 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:49]: Ok I will fix

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c (hunk 2 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:52]: OK

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+#define PORTFOLIO_IDENTITY_ID			0
+#define PORTFOLIO_GET_AUTH_DATA_ID		1
+#define PORTFOLIO_AUTH_DATA_ID			2
+#define PORTFOLIO_AUTH_STATUS_ID		3
+
+#define PORTFOLIO_MAX_ID 4
+
+/* Identity max based on LwM2M v1.1 conformance test requirements */
+#define PORTFOLIO_IDENTITY_MAX 4
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:31]: These seem to be usually Kconfig-options, don't know if it matters here as it is for conformance testing.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+#define PORTFOLIO_AUTH_DATA_MAX 4
+/* Support only 1 even this multi instance object */
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:33]: Comment seems to be mismatching with the define

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+static int lwm2m_portfolio_init(const struct device *dev)
+{
+	portfolio.obj_id = LWM2M_OBJECT_PORTFOLIO_ID;
+	portfolio.version_major = PORTFOLIO_VERSION_MAJOR;
+	portfolio.version_minor = PORTFOLIO_VERSION_MINOR;
+	portfolio.is_core = false;
+	portfolio.fields = fields;
+	portfolio.field_count = ARRAY_SIZE(fields);
+	portfolio.max_instance_count = ARRAY_SIZE(inst);
```
Comment (jarlamsa) [inline subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:111]: Why this is not using `MAX_INSTANCE_COUNT` instead?

**Review by jarlamsa — DISMISSED** [review]
LGTM, some nitpicks only

**Review by SeppoTakalo — COMMENTED** [review]
Look OK, but I would prefer to fix those remarks.

**Review by SeppoTakalo — APPROVED** [review]

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c (hunk 5 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:33]: Ah my bad I update comment.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c (hunk 6 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:111]: good point I use that one.

File: subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+ * add PORTFOLIO_IDENTITY_ID resource instances
+ * add PORTFOLIO_AUTH_DATA_ID resource instances
+ */
+#define RESOURCE_INSTANCE_COUNT        (PORTFOLIO_MAX_ID - 3 + \
+					PORTFOLIO_IDENTITY_MAX + \
+					PORTFOLIO_AUTH_DATA_MAX)
+
+static struct lwm2m_engine_obj portfolio;
+static struct lwm2m_engine_obj_field fields[] = {
+	OBJ_FIELD_DATA(PORTFOLIO_IDENTITY_ID, RW, STRING), /* Mandatory Multi Instance */
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_obj_portfolio.c:49]: Ok. I do that.

**Comment by jheiskan81** [comment]
@jarlamsa , @VeijoPesonen and @SeppoTakalo Thank you for feedback I fix you comment's.

**Review by SeppoTakalo — APPROVED** [review]

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

File: include/net/lwm2m.h [hunk 8]
```diff
@@ -45,6 +45,7 @@
 #define LWM2M_OBJECT_LOCATION_ID                6
 #define LWM2M_OBJECT_CONNECTIVITY_STATISTICS_ID 7
 #define LWM2M_OBJECT_SOFTWARE_MANAGEMENT_ID     9
+#define LWM2M_OBJECT_PORTFOLIO_ID		16
```
Comment (tejlmand) [inline include/net/lwm2m.h:48]: tabs are for indentation, not sub-alignment like here.

**Review by tejlmand — COMMENTED** [review]
minor nit

File: include/net/lwm2m.h (hunk 8 above)
Comment (jheiskan81) [reply include/net/lwm2m.h:48]: Ok, I will fix that.

**Review by SeppoTakalo — DISMISSED** [review]

**Review by VeijoPesonen — DISMISSED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**